docs/accelerator_configuration.md
# Accelerator Configuration

This file is a detailed explanation of the adjustable parameters. Most of these are found in [accelerator/device/params.hpp](../accelerator/device/params.hpp). 

Important: make sure the host-side software and the accelerator are compiled with the same settings!

General parameters:

- If MCBBS support is required, then define **SPW_SYSTEM**. Otherwise, define **DENSE_SYSTEM**. Do NOT define both.
- DDR_BANDWIDTH_GBS_INT: Estimation of the off-chip memory access bandwidth in GB/s. Used by the instruction generator to generate the best tile configuration.
- FMAX_MHZ: Clock frequency of the accelerator in MHz. Used by the instruction generator to generate the best tile configuration.
- PE_COLS: Number of columns of the convolution systolic array.
- PE_ROWS: Number of rows of the convolution systolic array.
- PE_ROW_GROUPS: PE column coalescing size, equivalent to the "G" parameter mentioned in the paper and the thesis.
- MISC_ACCUMULATOR_WIDTH: Number of bits allocated for each temporary value accumulator inside the MISC engine. Must be chosen from 32, 28, 24, 20, or 16.
- KERNEL_CACHE_SIZE_VALUE_BYTE: Number of bytes that each side of a weight buffer inside the CONV engine can hold.
- IA_CACHE_SIZE_BYTE: Number of bytes that each side of an input neuron buffer inside the CONV engine can hold.
- OA_CACHE_SIZE_BYTE: Number of bytes that each side of an output neuron buffer inside the CONV engine can hold.

Parameters that affect the PEs:

- CLUSTER_SIZE: MCBBS cluster size (C). 
- PRUNE_RANGE_IN_CLUSTER: Pruning range size in clusters (R).
- PE_SIMD_SIZE: Number of clusters that are multiplied in parallel (P). If SPW_SYSTEM is enabled, this is the processing window size.
- ACCUMULATOR_WIDTH: Number of bits allocated for the partial-sum accumulator in each PE. Must be chosen from 32, 28, 24, 20, or 16.

Parameters that access the off-chip data access block size:

- ACTIVATION_WIDE_SIZE: Number of neuron "fetch blocks" inside a neuron DRAM data block. If MCBBS is enabled, then each neuron fetch block contains P * C * R neurons. Otherwise, each neuron fetch block contains P * C neurons. This value also controls how many values are processed in parallel inside the MISC engine. **It must be an interger power of 2**.
- ACTIVATION_WIDE_SIZE_OFFSET: Set this to log2(ACTIVATION_WIDE_SIZE).
- ACTIVATION_WIDE_SIZE_REMAINDER_MASK: Set this to ACTIVATION_WIDE_SIZE - 1.
- WEIGHT_WIDE_SIZE: Number of weight "fetch blocks" inside a weight DRAM data block. The number of sparse/dense weights inside a weight fetch block is P * C. **It must be an interger power of 2**.
- WEIGHT_WIDE_SIZE_OFFSET: Set this to log2(WEIGHT_WIDE_SIZE).
- WEIGHT_WIDE_SIZE_REMAINDER_MASK: Set this to WEIGHT_WIDE_SIZE - 1.

During software emulation, it might be helpful to see some print statement outputs. To do so, please use the EMULATOR_PRINT macro inside [accelerator/device/prints.hpp](../accelerator/device/prints.hpp), and define the EMUPRINT flag.  

accelerator/device/params.hpp
#ifndef PARAMS_DEFINED
#define PARAMS_DEFINED

//#define HOST_DEBUG
//#define SPW_SYSTEM
#define DENSE_SYSTEM
#define OA_PING_PONG
//#define HW_SYNC

#define NOOP
#if defined(SPW_TEST)
#undef DENSE_SYSTEM
#define SPW_SYSTEM
#endif
/**
 * Global memory settings
 */
#define MAX_DRAM_BYTE_INPUT_ACTIVATION (1 << 29)
#define MAX_DRAM_BYTE_INPUT_ACTIVATION_SB_COUNT (1 << 22)
#define MAX_DRAM_BYTE_INPUT_WEIGHT (1 << 28)
#define MAX_DRAM_BYTE_INPUT_WEIGHT_SB_COUNT 0x40000
#define MAX_DRAM_BYTE_INPUT_BIAS 0x20000
//#define MAX_DRAM_BYTE_OUTPUT_ACTIVATION 0x400000
//#define MAX_DRAM_BYTE_OUTPUT_ACTIVATION_SB_COUNT 0x40000
#define MAX_DRAM_BYTE_INPUT_MOVER_INSTRUCTION (1 << 23)
#define MAX_DRAM_BYTE_INPUT_TILE_CONTROLLER_INSTRUCTION (1 << 23)
#define MAX_DRAM_BYTE_OUTPUT_MOVER_INSTRUCTION (1 << 23)
#define MAX_DRAM_BYTE_WEIGHT_MOVER_INSTRUCTION (1 << 25)
#define MAX_DRAM_BYTE_OUTPUT_TILE_CONTROLLER_INSTRUCTION (1 << 23)
#define MAX_DRAM_BYTE_MISC_CONTROLLER_INSTRUCTION (1 << 23)

// #define PACKET_SIZE 1

//Assume on Arria 10 Dev Kit, the memory bandwidth is on
#define DDR_BANDWIDTH_GBS_INT 17
#define FMAX_MHZ 221
#define DDR_BYTES_PER_CYCLE ((DDR_BANDWIDTH_GBS_INT * 1000 - 1) / FMAX_MHZ + 1)

#if defined(FULL_SYSTEM)
	// #define PE_COLS 2
	// #define PE_ROWS_PER_GROUP 4
	// #define PE_ROW_GROUPS 2
	// #define MISC_COLS 1
	#define PE_COLS 7
	#define PE_ROWS_PER_GROUP 8
	#define PE_ROW_GROUPS 3
	#define MISC_COLS 1
	// #define PE_COLS 1
	// #define PE_ROWS_PER_GROUP 1
	// #define PE_ROW_GROUPS 1
	// #define MISC_COLS 1
	// #define PE_COLS 7
	// #define PE_ROWS_PER_GROUP 8
	// #define PE_ROW_GROUPS 6
	// #define MISC_COLS 1
#else
	#define PE_COLS 1
	#define PE_ROWS_PER_GROUP 16
	#define PE_ROW_GROUPS 1
	#define MISC_COLS 1
#endif

#define PE_ROWS (PE_ROWS_PER_GROUP * PE_ROW_GROUPS)
//Derived parameter
#if (PE_ROWS_PER_GROUP == 1)
#define DIVIDE_BY_PE_ROWS_PER_GROUP_SHIFT 0
#define DIVIDE_BY_PE_ROWS_PER_GROUP_REMAINDER_MASK 0x0
#elif (PE_ROWS_PER_GROUP == 2)
#define DIVIDE_BY_PE_ROWS_PER_GROUP_SHIFT 1
#define DIVIDE_BY_PE_ROWS_PER_GROUP_REMAINDER_MASK 0x1
#elif (PE_ROWS_PER_GROUP == 4)
#define DIVIDE_BY_PE_ROWS_PER_GROUP_SHIFT 2
#define DIVIDE_BY_PE_ROWS_PER_GROUP_REMAINDER_MASK 0x3
#elif (PE_ROWS_PER_GROUP == 8)
#define DIVIDE_BY_PE_ROWS_PER_GROUP_SHIFT 3
#define DIVIDE_BY_PE_ROWS_PER_GROUP_REMAINDER_MASK 0x7
#elif (PE_ROWS_PER_GROUP == 16)
#define DIVIDE_BY_PE_ROWS_PER_GROUP_SHIFT 4
#define DIVIDE_BY_PE_ROWS_PER_GROUP_REMAINDER_MASK 0xF
#else
#error DIVIDE_BY_PE_ROWS_PER_GROUP_SHIFT should be chosen from {1, 2, 4, 8, 16}
#endif

#if (MISC_COLS > PE_COLS)
#error Configuration MISC_COLS should be less or equal to PE_COLS
#endif

#define CHANNEL_DEPTH 1
#define OA_DRAIN_CHANNEL_DEPTH 1

//Activation memory region offsets
//in terms of DRAM blocks
//TB count memory region offsets
//In terms of TB counts (shorts)
#define NUM_ACTIVATION_REGIONS 3
#define MEM_ACTIVATION_REGION_SIZE_PER_SLICE (1 << 20)
#define MEM_ACTIVATION_TB_REGION_SIZE_PER_SLICE (1 << 18)


#define CLUSTER_SIZE 2 //cluster size in terms of values

#if (CLUSTER_SIZE == 1)
#define VALUE_TO_CLUSTER_SHIFT 0x00
#elif (CLUSTER_SIZE == 2)
#define VALUE_TO_CLUSTER_SHIFT 0x01
#elif (CLUSTER_SIZE == 4)
#define VALUE_TO_CLUSTER_SHIFT 0x02
#elif (CLUSTER_SIZE == 8)
#define VALUE_TO_CLUSTER_SHIFT 0x03
#elif (CLUSTER_SIZE == 16)
#define VALUE_TO_CLUSTER_SHIFT 0x04
#else
#error CLUSTER_SIZE should be chosen from {1, 2, 4, 8, 16}
#endif 
#define VALUE_DIVIDED_BY_CLUSTER_SIZE_REMAINDER_MASK ((1 << VALUE_TO_CLUSTER_SHIFT) - 1)


#define VALUE_DIVIDED_BY_SIMD_SIZE_REMAINDER_MASK ((1 << (VALUE_TO_CLUSTER_SHIFT + CLUSTER_TO_TRANSFER_SIZE_SHIFT)) - 1)
#define CLUSTER_TO_TRANSFER_BLOCK_SHIFT CLUSTER_TO_TRANSFER_SIZE_SHIFT //amount of right shift required to convert a cluster count into transfer block count

/**
 * Parameters relevant for balanced-block sparsity
 */
//Prune range of balanced-sparsity
#define PRUNE_RANGE_IN_CLUSTER 4
//Number of prune range processed in parallel.
//Equal to SIMD size (in terms of cluster) in a SpW PE
#define PE_SIMD_SIZE 8

#if (PE_SIMD_SIZE == 1)
#define PE_SIMD_SIZE_CLUSTER_OFFSET 0x0
#elif (PE_SIMD_SIZE == 2)
#define PE_SIMD_SIZE_CLUSTER_OFFSET 0x01
#elif (PE_SIMD_SIZE == 4)
#define PE_SIMD_SIZE_CLUSTER_OFFSET 0x02
#elif (PE_SIMD_SIZE == 8)
#define PE_SIMD_SIZE_CLUSTER_OFFSET 0x03
#elif (PE_SIMD_SIZE == 16)
#define PE_SIMD_SIZE_CLUSTER_OFFSET 0x04
#else
#error PE_SIMD_SIZE should be chosen from {1, 2, 4, 8, 16}
#endif
#define PE_SIMD_SIZE_CLUSTER_MASK ((1 << (PE_SIMD_SIZE_CLUSTER_OFFSET)) - 1)

#if (PRUNE_RANGE_IN_CLUSTER == 1)
#define PRUNE_RNAGE_IN_CLUSTER_OFFSET 0x0
#elif (PRUNE_RANGE_IN_CLUSTER == 2)
#define PRUNE_RNAGE_IN_CLUSTER_OFFSET 0x01
#elif (PRUNE_RANGE_IN_CLUSTER == 4)
#define PRUNE_RNAGE_IN_CLUSTER_OFFSET 0x02
#elif (PRUNE_RANGE_IN_CLUSTER == 8)
#define PRUNE_RNAGE_IN_CLUSTER_OFFSET 0x03
#elif (PRUNE_RANGE_IN_CLUSTER == 16)
#define PRUNE_RNAGE_IN_CLUSTER_OFFSET 0x04
#else
#error PRUNE_RANGE_IN_CLUSTER should be chosen from {1, 2, 4, 8, 16}
#endif
#define PRUNE_RANGE_IN_CLUSTER_MASK ((1 << (PRUNE_RNAGE_IN_CLUSTER_OFFSET)) - 1)

//Size of the char array in the host weight blocks
//used to contain the indices of the NZ clusters
//Each char is split into two 4-bit halfs.
//Each half corresponds to an index
#if defined(PE_SIMD_SIZE) && defined(SPW_SYSTEM)
#if (PE_SIMD_SIZE <= 2)
#define INDEX_CHAR_ARRAY_SIZE 1
#define INDEX_CHAR_ARRAY_SIZE_OFFSET 0
#elif (PE_SIMD_SIZE <= 4)
#define INDEX_CHAR_ARRAY_SIZE 2
#define INDEX_CHAR_ARRAY_SIZE_OFFSET 1
#elif (PE_SIMD_SIZE <= 8)
#define INDEX_CHAR_ARRAY_SIZE 4
#define INDEX_CHAR_ARRAY_SIZE_OFFSET 2
#elif (PE_SIMD_SIZE <= 16)
#define INDEX_CHAR_ARRAY_SIZE 8
#define INDEX_CHAR_ARRAY_SIZE_OFFSET 3
#else
#error PE_SIMD_SIZE should be chosen from {1, 2, 4, 8, 16}
#endif
#endif

#if defined(SPW_SYSTEM)
	#define PE_ACTIVATION_BLOCK_SIZE_IN_WORD (CLUSTER_SIZE * PE_SIMD_SIZE * PRUNE_RANGE_IN_CLUSTER)
	#define PE_ACTIVATION_BLOCK_SIZE_IN_WORD_OFFSET (VALUE_TO_CLUSTER_SHIFT + PE_SIMD_SIZE_CLUSTER_OFFSET + PRUNE_RNAGE_IN_CLUSTER_OFFSET)
#else
	#define PE_ACTIVATION_BLOCK_SIZE_IN_WORD (CLUSTER_SIZE * PE_SIMD_SIZE)
	#define PE_ACTIVATION_BLOCK_SIZE_IN_WORD_OFFSET (VALUE_TO_CLUSTER_SHIFT + PE_SIMD_SIZE_CLUSTER_OFFSET)
#endif
#define PE_ACTIVATION_BLOCK_SIZE_IN_WORD_MASK ((1 << PE_ACTIVATION_BLOCK_SIZE_IN_WORD_OFFSET) - 1)


#define SURVIVING_COUNT_CLUSTER_INDEX 0X1
#define SURVIVING_COUNT_TRANSFER_BLOCK_INDEX 0x1

#define ACTIVATION_WIDE_SIZE 4
#define ACTIVATION_WIDE_SIZE_OFFSET 0x2 //Numnber of bits to shift the transfer block index to the right in order to recover the wide offset
#define ACTIVATION_WIDE_SIZE_REMAINDER_MASK 0x3
#define ACTIVATION_BURST_SIZE_BYTE (PE_ACTIVATION_BLOCK_SIZE_IN_WORD * ACTIVATION_WIDE_SIZE)
#define ACTIVATION_BURST_SIZE_BYTE_OFFSET (ACTIVATION_WIDE_SIZE_OFFSET + PE_ACTIVATION_BLOCK_SIZE_IN_WORD_OFFSET)
#define ACTIVATION_WIDE_SIZE_BYTE_MASK ((1 << ACTIVATION_BURST_SIZE_BYTE_OFFSET) - 1)
#define ACTIVATION_BURST_SIZE_IN_PE_ROW_GROUP (ACTIVATION_BURST_SIZE_BYTE / PE_ROWS_PER_GROUP)
#define ACTIVATION_BURST_SIZE_IN_PE_ROW_GROUP_OFFSET (ACTIVATION_BURST_SIZE_BYTE_OFFSET - DIVIDE_BY_PE_ROWS_PER_GROUP_SHIFT)
#if (ACTIVATION_BURST_SIZE_IN_PE_ROW_GROUP_OFFSET < 0)
#error The number of rows in a PE group must be smaller or equal to the activation burst size
#endif
#define ACTIVATION_WIDE_SIZE_IN_PE_ROW_GROUP_MASK ((1 << ACTIVATION_BURST_SIZE_IN_PE_ROW_GROUP_OFFSET) - 1)

#define WEIGHT_WIDE_SIZE 4
#define WEIGHT_WIDE_SIZE_OFFSET 0x2 //Numnber of bits to shift the transfer block index to the right in order to recover the wide offset
#define WEIGHT_WIDE_SIZE_REMAINDER_MASK 0x3
#define WEIGHT_BURST_SIZE_VALUE_BYTE (WEIGHT_WIDE_SIZE * PE_SIMD_SIZE * CLUSTER_SIZE)
#if defined(SPW_SYSTEM)
#define WEIGHT_BURST_SIZE_INDEX_BYTE (WEIGHT_WIDE_SIZE*INDEX_CHAR_ARRAY_SIZE)
#endif


#define WMOVER_FILTER_DRAM_BLOCK_ACCESS_UNROLL_FACTOR 4
#define KERNEL_CACHE_SIZE_VALUE_BYTE 32768
#define KERNEL_CACHE_DEPTH (KERNEL_CACHE_SIZE_VALUE_BYTE / WEIGHT_BURST_SIZE_VALUE_BYTE)

#define MAX_OUTPUT_TILE_WIDTH_PER_COL 8 
#define MAX_OUTPUT_TILE_HEIGHT 31
#define MAX_INPUT_TILE_WIDTH_PER_COL 8
#define MAX_INPUT_TILE_HEIGHT 31

// #define IA_CACHE_SIZE_BYTE 16384
#define IA_CACHE_SIZE_BYTE 32768
#define IA_CACHE_DEPTH (IA_CACHE_SIZE_BYTE/ACTIVATION_BURST_SIZE_BYTE)

#define WEIGHT_MOVER_BIAS_CACHE_SIZE 2048

//TODO: Change this back to the commented line
//#define OA_CACHE_SIZE_BYTE 8192
#define OA_CACHE_SIZE_BYTE 32768
#define OA_CACHE_DEPTH (OA_CACHE_SIZE_BYTE/ACTIVATION_BURST_SIZE_BYTE)

//Accumulator width
#define ACCUMULATOR_WIDTH 28
#if defined(EMULATOR)
#pragma message("WARNING: IN EMULATOR MODE, ACCUMULATOR_WIDTH IS FIXED TO 32")
#define ACCUM_MASK 0x0FFFFFFFF
#define MULT_MASK 0x0FFFFFFFF
#define ACCUM_MIN 0x80000000
#elif (ACCUMULATOR_WIDTH == 32)
#define ACCUM_MASK 0x0FFFFFFFF
#define MULT_MASK 0x0FFFFFFFF
#define ACCUM_MIN 0x80000000
#elif (ACCUMULATOR_WIDTH == 28)
#define ACCUM_MASK 0x00FFFFFFF
#define MULT_MASK 0x00FFFFFFF
#define ACCUM_MIN 0x08000000
#elif (ACCUMULATOR_WIDTH == 24)
#define ACCUM_MASK 0x00FFFFFF
#define MULT_MASK 0x00FFFFFF
#define ACCUM_MIN 0x00800000
#elif (ACCUMULATOR_WIDTH == 20)
#define ACCUM_MASK 0x000FFFFF
#define MULT_MASK 0x000FFFFF
#define ACCUM_MIN 0x00080000
#elif (ACCUMULATOR_WIDTH == 16)
#define ACCUM_MASK 0x0FFFF
#define MULT_MASK 0x0FFFF
#define ACCUM_MIN 0x00008000
#else
#error Accumulator width should be from 32-bit, 28-bit, 24-bit, 20-bit, and 16-bit
#endif

#define MISC_ACCUMULATOR_WIDTH 16
#if defined(EMULATOR)
#pragma message("WARNING: IN EMULATOR MODE, MISC_ACCUMULATOR_WIDTH IS FIXED TO 32")
#define MISC_ACCUM_MASK 0x0FFFFFFFF
#define MISC_MULT_MASK 0x0FFFFFFFF
#define MISC_ACCUM_MIN 0x80000000
#elif (MISC_ACCUMULATOR_WIDTH == 32)
#define MISC_ACCUM_MASK 0x0FFFFFFFF
#define MISC_MULT_MASK 0x0FFFFFFFF
#define MISC_ACCUM_MIN 0x80000000
#elif (MISC_ACCUMULATOR_WIDTH == 28)
#define MISC_ACCUM_MASK 0x00FFFFFFF
#define MISC_MULT_MASK 0x00FFFFFFF
#define MISC_ACCUM_MIN 0x08000000
#elif (MISC_ACCUMULATOR_WIDTH == 24)
#define MISC_ACCUM_MASK 0x00FFFFFF
#define MISC_MULT_MASK 0x00FFFFFF
#define MISC_ACCUM_MIN 0x00800000
#elif (MISC_ACCUMULATOR_WIDTH == 20)
#define MISC_ACCUM_MASK 0x000FFFFF
#define MISC_MULT_MASK 0x000FFFFF
#define MISC_ACCUM_MIN 0x00080000
#elif (MISC_ACCUMULATOR_WIDTH == 16)
#define MISC_ACCUM_MASK 0x0FFFF
#define MISC_MULT_MASK 0x0FFFF
#define MISC_ACCUM_MIN 0x00008000
#else
#error Misc accumulator width should be from 32-bit, 28-bit,  24-bit, 20-bit, and 16-bit
#endif


#define TIMEOUT 0X1FFFFFF
#endif

